lib/pdn/pdn.c

```diff
@@ -0,0 +1,457 @@
+	if (p) {
+		esm_err = atoi(p + strlen("CNEC_ESM: "));
+
+		p = strchr(p, ',');
+		if (!p) {
+			return;
+		}
+
+		cid = atoi(p + 1);
+
+		pdn = pdn_find(cid);
+		if (!pdn) {
+			return;
+		}
+
+		if (pdn->callback) {
+			pdn->callback((intptr_t)cid, PDN_EVENT_CNEC_ESM,
+				      esm_err);
+		}
+
+		esm_from_notif = esm_err;
+		k_sem_give(&notif_sem);
+
+		return;
+	}
+
+	const struct {
+		const char *notif;
+		const enum pdn_event event;
+	} map[] = {
+		{"CGEV: ME PDN ACT",	PDN_EVENT_ACTIVATED},	/* +CGEV: ME PDN ACT <cid> */
+		{"CGEV: ME PDN DEACT",	PDN_EVENT_DEACTIVATED},	/* +CGEV: ME PDN DEACT <cid> */
+		/* Order is important */
+		{"CGEV: IPV6 FAIL",	PDN_EVENT_IPV6_DOWN},	/* +CGEV: IPV6 FAIL <cid> */
+		{"CGEV: IPV6",		PDN_EVENT_IPV6_UP},	/* +CGEV: IPV6 <cid> */
+	};
+
+	for (size_t i = 0; i < ARRAY_SIZE(map); i++) {
+		p = strstr(notif, map[i].notif);
+		if (!p) {
+			continue;
+		}
+
+		cid = atoi(p + strlen(map[i].notif));
+
+		if (default_callback && cid == 0) {
+			default_callback(0, map[i].event, 0);
+			return;
+		}
+
+		pdn = pdn_find(cid);
+		if (!pdn) {
+			return;
+		}
+
+		if (pdn->callback) {
+			pdn->callback((intptr_t)cid, map[i].event, 0);
+			return;
+		}
+	}
+}
+
+int pdn_init(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(pdn_contexts); i++) {
+		pdn_contexts[i].context_id = CID_UNASSIGNED;
+	}
+
+	k_sem_init(&notif_sem, 0, 1);
+
+#if defined(CONFIG_PDN_LEGACY_PCO)
+	err = at_cmd_write("AT%XEPCO=0", NULL, 0, NULL);
+	if (err) {
+		LOG_WRN("Failed to set legacy PCO mode, err %d", err);
+	}
+#endif
+
+	return at_notif_register_handler(NULL, at_notif_handler);
+}
+
+int pdn_default_callback_set(pdn_event_handler_t cb)
+{
+	if (!cb) {
+		return -EINVAL;
+	}
+
+	default_callback = cb;
+
+	return 0;
+}
+
+int pdn_cid_create(uint8_t *cid, pdn_event_handler_t cb)
+{
+	int err;
+	char *p;
+	struct pdn *pdn;
+
+	if (!cid) {
+		return -EINVAL;
+	}
+
+	pdn = pdn_avail_get();
+	if (!pdn) {
+		return -ENOBUFS;
+	}
+
+	err = at_cmd_write("AT%XNEWCID?", at_buf, sizeof(at_buf), NULL);
+	if (err) {
+		return err;
+	}
+
+	p = strchr(at_buf, ':');
+	if (!p) {
+		return -EBADMSG;
+	}
+
+	pdn->context_id = atoi(p + 1);
+
+	*cid = pdn->context_id;
+	if (cb) {
+		pdn->callback = cb;
+	}
+
+	return 0;
+}
+
+int pdn_cid_configure(uint8_t cid, const char *apn, enum pdn_fam fam,
+		      struct pdn_pdp_opt *opt)
+{
+	int err;
+	char family[] = "IPV4V6"; /* longest family */
+
+	switch (fam) {
+	case PDN_FAM_IPV4:
+		strncpy(family, "IP", sizeof(family));
+		break;
+	case PDN_FAM_IPV6:
+		strncpy(family, "IPV6", sizeof(family));
+		break;
+	case PDN_FAM_IPV4V6:
+		strncpy(family, "IPV4V6", sizeof(family));
+		break;
+	case PDN_FAM_NONIP:
+		strncpy(family, "Non-IP", sizeof(family));
+		break;
+	}
+
+	if (opt) {
+		/* NOTE: this is a workaround for an issue
+		 * where passing default values does not work as intended.
+		 * Only format these params when `opt` is passed explicitly.
+		 */
+		err = snprintf(at_buf, sizeof(at_buf),
```

Comment:
Is there a reason for using `snprintf` over `snprintk`? Wonder if we can get away with a smaller footprint with `snprintk`

Reply:
I did not know about `snprintk` :) Have not looked too deep into the differences, but looking at this:
```
snprintf
           FLASH:       62248 B       960 KB      6.33%
```
```
snprintk
           FLASH:       62344 B       960 KB      6.34%
```
I think we can stick `snprintf` here.

Reply:
That was a bit surprising, I agree that it's best to stick with `snprintf`